### ISSUE TITLE
Printing is deferred in RStudio

### ISSUE THREAD

**Issue by hafen** [rstudio/reticulate#685]
If I run the following code in RStudio:

```r
library(reticulate)
builtins <- import_builtins()
builtins$print("foo")
```

I get no output. However, if I run it from an R terminal, I get the expected output:

```
foo
```

Note that in RStudio, if I later trigger anything in Python with reticulate (even running an empty Python command)

```r
library(reticulate)
builtins <- import_builtins()
builtins$print("foo")
# do some other stuff...
py_run_string("")
```

I will finally see the printed output:

```
foo
```

This is a critical issue in that for Python packages that print messages for the user, a user using RStudio will likely never see the messages, and if they do, it will not be at the time the message was intended to be printed.

**Comment by kevinushey** [comment]
The issue can be worked around by explicitly flushing the connection:

```
> builtins$print("foo", flush = TRUE)
foo
```

We might want to consider auto-flushing or something similar.

**Comment by hafen** [comment]
Thanks. Since posting I also realized that this is Python-specific. It flushes in RStudio using my system-installed Python but not with anaconda.

In my case, `flush = TRUE` isn't possible because I am wrapping a Python library that I don't have control over. But I realized I can set

```r
Sys.setenv(PYTHONUNBUFFERED = TRUE)
```

as a workaround.

I'm wondering what the difference is between RStudio and terminal R that causes the difference in behavior.

**Comment by kevinushey** [comment]
I've made the following change to auto-flush output when R reaches the top level:

https://github.com/rstudio/reticulate/commit/72f3dceb2c5280ac5ce7f18b4fd78872cf4e583c

Does this suffice? Or do you think ensuring unbuffered output is better here?

**Comment by hafen** [comment]
Yes I was thinking of something like this. Thanks! I suppose the only additional thing would be if you want to honor the `PYTHONUNBUFFERED` environment variable if someone has explicitly set it.

**Comment by kevinushey** [comment]
Good idea!
